Testing.Rmd

```diff
@@ -442,159 +769,1106 @@ It's hard to give good general advice about writing tests, but you might find th
+Don't worry, you're not!
+Testing presents many unique challenges and maneuvers, which tend to get much less air time in programming communities than strategies for writing the "main code", i.e. the stuff below `R/`.
+As a result, it requires more deliberate effort to develop your skills and taste around testing.
+
+Anecdotally, it seems like many new R package developers fall back to habits from data analysis code, when writing tests.
+To a certain extent, this is natural, since your tests need to exercise the functions in your package by doing real tasks.
+The code *inside each `test_that()` call* should be very procedural.
+But at the next level of organisation, i.e. a file with multiple tests or a folder with many test files, the step-by-step script mentality can lead you astray.
+
+Recall this advice found in section \@ref(code-r-landscape), which covers your package's "main code", i.e. everything below `R/`:
+
+> The `.R` files below `R/` should consist almost entirely of function definitions.
+> Any other top-level code is suspicious and should be carefully reviewed for possible conversion into a function.
+
+We have analogous advice for your "testing code", i.e. everything in `tests/testthat/test-*.R` files:
+
+> The `test-*.R` file below `tests/testthat/` should consist almost entirely of calls to `test_that()`.
+> Any other top-level code is suspicious and should be carefully reviewed for conversion to a more official method of achieving its goal, such as using functions or data objects defined in your package or relocating logic to special testthat files, such as `helper.R` or `setup.R`.
 
 ```{r, eval = FALSE}
-check_api <- function() {
-  if (not_working()) {
-    skip("API not available")
-  }
-}
+# sometimes people put top-level code here, but it's a good idea to avoid this
+# when you can
 
-test_that("foo api returns bar when given baz", {
-  check_api()
-  ...
-})
+test_that("foofy() does this", { ... })
+
+test_that("foofy() does that", { ... })
+
+# it's even more problematic to have top-level code in the MIDDLE of a test file
+#
+# such code only affects *part* of your test file
+#
+# this code is hard to discover and run when troubleshooting
+
+test_that("foofy() does the other thing", { ... })
 ```
 
-### Building your own testing tools
+Keep reading for various alternatives that are better than having top-level code in your test files other than calls to `test_that()`.
 
-As you start to write more tests, you might notice duplication in your code.
-For example, the following code shows one test of the `floor_date()` function from `library(lubridate)`.
-There are seven expectations that check the results of rounding a date down to the nearest second, minute, hour, etc.
-There's a lot of duplication (which increases the chance of bugs), so we might want to extract common behaviour into a new function.
+### Test hygiene
```

Comment:
One question here! Also related to refactoring code.

I somewhat often use the pattern

```r
is_bla <- function() {
 if (nzchar(Sys.getenv("PACKAGE_BLA_MOCK")) {
    return(FALSE)
}
  # actual code checking bla
}

```

to have a switch for testing for things such as installed packages, internet connection. Is this... good? bad?